File: IntegrationTests/BraintreeApplePay_IntegrationTests.m
#import "BTNonceValidationHelper.h"
#import <PassKit/PassKit.h>
#import <XCTest/XCTest.h>
#import <IntegrationTests-Swift.h>

@import BraintreeCore;
@import BraintreeApplePay;

@interface BraintreeApplePay_IntegrationTests : XCTestCase

@end

@implementation BraintreeApplePay_IntegrationTests

- (void)testTokenizeApplePayPayment_whenApplePayEnabledInControlPanel_returnsANonce {
    BTAPIClient *apiClient = [[BTAPIClient alloc] initWithAuthorization:BTIntegrationTestsConstants.sandboxTokenizationKey];
    BTApplePayClient *client = [[BTApplePayClient alloc] initWithAPIClient:apiClient];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Tokenize Apple Pay payment"];
    [client tokenizeApplePayPayment:[[PKPayment alloc] init]
                         completion:^(BTApplePayCardNonce * _Nullable tokenizedApplePayPayment, NSError * _Nullable error) {
        XCTAssertTrue(tokenizedApplePayPayment.nonce.isANonce);
        XCTAssertNil(error);
        [expectation fulfill];
    }];

    [self waitForExpectationsWithTimeout:5 handler:nil];
}

- (void)testTokenizeApplePayPayment_whenApplePayDisabledInControlPanel_returnsError {
    BTAPIClient *apiClient = [[BTAPIClient alloc] initWithAuthorization:SANDBOX_TOKENIZATION_KEY_APPLE_PAY_DISABLED];
    BTApplePayClient *client = [[BTApplePayClient alloc] initWithAPIClient:apiClient];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Tokenize Apple Pay payment"];
    [client tokenizeApplePayPayment:[[PKPayment alloc] init]
                         completion:^(BTApplePayCardNonce * _Nullable tokenizedApplePayPayment, NSError * _Nullable error) {
        XCTAssertEqualObjects(error.domain, @"com.braintreepayments.BTApplePayErrorDomain");
        XCTAssertEqual(error.code, 1);
        XCTAssertNil(tokenizedApplePayPayment);
        [expectation fulfill];
    }];

    [self waitForExpectationsWithTimeout:5 handler:nil];
}

@end

File: IntegrationTests/Braintree-API-Integration-Specs/BraintreeApplePay_IntegrationTests.swift
import XCTest
import PassKit
@testable import BraintreeCore
@testable import BraintreeApplePay

class BraintreeApplePay_IntegrationTests: XCTestCase {

    func testTokenizeApplePayPayment_whenApplePayEnabledInControlPanel_returnsANonce() {
        let apiClient = BTAPIClient(authorization: BTIntegrationTestsConstants.sandboxTokenizationKey)!
        let applePayClient = BTApplePayClient(apiClient: apiClient)
        let expectation = expectation(description: "Tokenize Apple Pay payment")

        applePayClient.tokenize(PKPayment()) { tokenizedApplePayAccount, error in
            guard let nonce = tokenizedApplePayAccount?.nonce else {
                XCTFail("Nonce expected to be returned in this tests")
                return
            }

            XCTAssertTrue(nonce.isValidNonce)
            XCTAssertNil(error)
            expectation.fulfill()
        }

        waitForExpectations(timeout: 5)
    }

    func testTokenizeApplePayPayment_whenApplePayDisabledInControlPanel_returnsError() {
        let apiClient = BTAPIClient(authorization: BTIntegrationTestsConstants.sandboxTokenizationKeyApplePayDisabled)!
        let applePayClient = BTApplePayClient(apiClient: apiClient)
        let expectation = expectation(description: "Tokenize Apple Pay payment")

        applePayClient.tokenize(PKPayment()) { nonce, error in
            guard let error = error as? NSError else {
                XCTFail("Error expected to be returned in this tests")
                return
            }

            XCTAssertNil(nonce)
            XCTAssertEqual(error.domain, BTApplePayError.errorDomain)
            XCTAssertEqual(error.code, BTApplePayError.unsupported.errorCode)
            expectation.fulfill()
        }

        waitForExpectations(timeout: 5)
    }
}
